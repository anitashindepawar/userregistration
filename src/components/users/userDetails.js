import React from 'react';
import { connect } from 'react-redux';
import { getUserByUserID, createUser, updateUser, resetUserState } from '../../actionCreator/users'


export class UserDetails extends React.Component {
    constructor(props) {
        super(props);
        this.state = { userName: '', companyName: '', roleName: '' };
    }

    componentDidMount() {
        const userId = this.props.match.params.id;
        if (userId) {
            this.props.getUserByUserID(userId);
        } //else {
        //     this.setState()
        // }
    }

    handleChange = (e) => {
        const name = e.target.id;
        const value = e.target.value;
        this.setState({ [name]: value });
    }

    renderUserDetails() {
        const { userDet } = this.props;
        return (<form className="ui form">
            <div className="field">
                <label>User</label>

                <input type="text" id="userName" name="userName"
                    placeholder="User Name" onChange={this.handleChange}
                    value={this.state.userName || userDet.name} />
            </div>
            <div className="field">
                <label>Company Name</label>
                <input type="text" id="companyName" name="companyName"
                    placeholder="Company Name" onChange={this.handleChange}
                    value={this.state.companyName || userDet.company} />
            </div>
            <div className="field">
                <label>Role</label>
                <input type="text" name="roleName" placeholder="Role" id="roleName"
                    onChange={this.handleChange}
                    value={this.state.roleName || userDet.role} />
            </div>
            <div>
                <button className="ui button" id = "submitButton" type="submit" onClick={this.submitUserDetails}>Submit</button>
                <button className="ui button" id = "cancelButton" type="button" onClick={this.cancelUpdation}>Cancel</button>
            </div>
        </form>);
    };


    render() {
        return (this.renderUserDetails());
    };

    componentDidUpdate() {

        if (this.props.isUserCreated || this.props.isUserUpdated) {
            this.props.resetUserState();
            this.props.history.push('/');
        }
    };

    cancelUpdation = () => {
        this.props.resetUserState();
        this.props.history && 
        this.props.history.push('/');
    }

    submitUserDetails = (e) => {
        e.preventDefault();
        const { userDet } = this.props;
        const newUser = {
            name: this.state.userName || userDet.name,
            role: this.state.roleName || userDet.role,
            company: this.state.companyName || userDet.company
        }

        if (this.props.location && this.props.location.pathname.includes('create')) {
            this.props.createUser(newUser);
        } else {
            newUser.id = this.props.match.params.id;
            this.props.updateUser(newUser);
        }

    };

};

const mapStateToProps = (state) => {
    return {
        userDet: state.users.fetchedUser,
        isUserCreated: state.users.isUserCreated,
        isUserUpdated: state.users.isUserUpdated
    };
}
export default connect(mapStateToProps, { getUserByUserID, createUser, updateUser, resetUserState })(UserDetails);