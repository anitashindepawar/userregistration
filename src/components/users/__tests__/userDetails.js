import Enzyme, { shallow, mount } from 'enzyme';
import React from "React";
import { UserDetails } from '../userDetails';


import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() })

describe("Testing UserDetails component", () => {

    const prop1 = {
        name: "abc",
        role: "role",
        company: "comp"

    }
    const userFunction = jest.fn();

    it("Test form and snapshot", () => {
        const wrapper = shallow(<UserDetails
            userDet={prop1}
            match={{ params: { id: 1 } }}
            getUserByUserID={userFunction} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("form")).toHaveLength(1);

    })

    it("Test mount and props", () => {
        const wrapper = mount(<UserDetails
            userDet={prop1}
            match={{ params: { id: 1 } }}
            getUserByUserID={userFunction} />);
        //expect(wrapper).toMatchSnapshot();
        expect(wrapper.props().match).toEqual({ "params": { "id": 1 } });
        expect(wrapper.props().userDet).toEqual(prop1);

    })

    it("Test button click", () => {
        const wrapper = mount(<UserDetails
            userDet={prop1}
            match={{ params: { id: 1 } }}
            getUserByUserID={userFunction}
            updateUser={userFunction}
        />);
        wrapper.find('#submitButton').simulate('click');
        expect(userFunction).toHaveBeenCalled();

    })

    it("Test cancel button click", () => {
        const wrapper = mount(<UserDetails
            userDet={prop1}
            match={{ params: { id: 1 } }}
            getUserByUserID={userFunction}
            updateUser={userFunction}
            resetUserState={userFunction}
        />);
        wrapper.find('#cancelButton').simulate('click');
        expect(userFunction).toHaveBeenCalled();

    })
})