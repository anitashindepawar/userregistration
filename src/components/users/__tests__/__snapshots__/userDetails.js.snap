// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UserDetails component Test form 1`] = `ShallowWrapper {}`;

exports[`Testing UserDetails component Test form and snapshot 1`] = `ShallowWrapper {}`;

exports[`Testing UserDetails component Test mount and props 1`] = `ShallowWrapper {}`;
